src/front/wgsl/ast.rs

```diff
@@ -0,0 +1,344 @@
+pub struct FunctionArgument<'a> {
+    pub name: Ident<'a>,
+    pub ty: Type<'a>,
+    pub binding: Option<crate::Binding>,
+    pub handle: Handle<Local>,
+}
+
+#[derive(Debug)]
+pub struct FunctionResult<'a> {
+    pub ty: Type<'a>,
+    pub binding: Option<crate::Binding>,
+}
+
+#[derive(Debug)]
+pub struct EntryPoint {
+    pub stage: crate::ShaderStage,
+    pub early_depth_test: Option<crate::EarlyDepthTest>,
+    pub workgroup_size: [u32; 3],
+}
+
+#[derive(Debug)]
+pub struct Function<'a> {
+    pub entry_point: Option<EntryPoint>,
+    pub name: Ident<'a>,
+    pub arguments: Vec<FunctionArgument<'a>>,
+    pub result: Option<FunctionResult<'a>>,
+    pub expressions: Arena<Expression<'a>>,
+    pub locals: Arena<Local>,
+    pub body: Block<'a>,
+}
+
+#[derive(Debug)]
+pub struct GlobalVariable<'a> {
+    pub name: Ident<'a>,
+    pub space: crate::AddressSpace,
+    pub binding: Option<crate::ResourceBinding>,
+    pub ty: Type<'a>,
+    pub init: Option<Handle<Expression<'a>>>,
+}
+
+#[derive(Debug)]
+pub struct StructMember<'a> {
+    pub name: Ident<'a>,
+    pub ty: Type<'a>,
+    pub binding: Option<crate::Binding>,
+    pub align: Option<(u32, Span)>,
+    pub size: Option<(u32, Span)>,
+}
+
+#[derive(Debug)]
+pub struct Struct<'a> {
+    pub name: Ident<'a>,
+    pub members: Vec<StructMember<'a>>,
+}
+
+#[derive(Debug)]
+pub struct TypeAlias<'a> {
+    pub name: Ident<'a>,
+    pub ty: Type<'a>,
+}
+
+#[derive(Debug)]
+pub struct Const<'a> {
+    pub name: Ident<'a>,
+    pub ty: Option<Type<'a>>,
+    pub init: Handle<Expression<'a>>,
+}
+
+#[derive(Debug)]
+pub struct Type<'a> {
+    pub kind: TypeKind<'a>,
+    pub span: Span,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum ArraySize<'a> {
+    Constant(Handle<Expression<'a>>),
+    Dynamic,
+}
+
+#[derive(Debug)]
+pub enum TypeKind<'a> {
+    Scalar {
+        kind: crate::ScalarKind,
+        width: crate::Bytes,
+    },
+    Vector {
+        size: crate::VectorSize,
+        kind: crate::ScalarKind,
+        width: crate::Bytes,
+    },
+    Matrix {
+        columns: crate::VectorSize,
+        rows: crate::VectorSize,
+        width: crate::Bytes,
+    },
+    Atomic {
+        kind: crate::ScalarKind,
+        width: crate::Bytes,
+    },
+    Pointer {
+        base: Box<Type<'a>>,
+        space: crate::AddressSpace,
+    },
+    Array {
+        base: Box<Type<'a>>,
+        size: ArraySize<'a>,
+    },
+    Image {
+        dim: crate::ImageDimension,
+        arrayed: bool,
+        class: crate::ImageClass,
+    },
+    Sampler {
+        comparison: bool,
+    },
+    BindingArray {
+        base: Box<Type<'a>>,
+        size: ArraySize<'a>,
+    },
+    User(Ident<'a>),
+}
+
+#[derive(Debug, Default)]
+pub struct Block<'a> {
+    pub stmts: Vec<Statement<'a>>,
+}
+
+#[derive(Debug)]
+pub struct Statement<'a> {
+    pub kind: StatementKind<'a>,
+    pub span: Span,
+}
+
+#[derive(Debug)]
+pub enum StatementKind<'a> {
+    VarDecl(Box<VarDecl<'a>>),
```

Comment:
I think this box can be removed.

Reply:
Also, let's rename the variant and type to `LocalDecl` (similar to our `GlobalDecl`).

Reply:
Without that, the size of a statement is severely bloated.

Performance doesn't really change anyways, even with our relatively declaration-heavy tests.

Reply:
Hmm, 56 vs 80.

But if we put `ast::Type` in an `Arena` and share `Handle`s in the AST, it gets us back down to 56.

Reply:
I still think that pessimising the 'normal' path (non-nested types) to arena pushes won't be worth it compared to the relatively rare nested types and local declaration.

Though, as is usual with most micro-optimizations like this, I really doubt it's gonna make any difference that isn't just normal benchmark noise.

Reply:
I tried this change locally and got 2-3% better perf (compared with this branch).
It's not groundbreaking but it allows us to still be performant with arbitrary nesting while also keeping things consistent.

Reply:
Arenas are the way to go here.